### PR TITLE
[WIP] support for HT1632C LED control

### PR DESCRIPTION
**Pull request by ericandrewlewis**
This is a work in progress PR for #1000. More or less a JS port of [Adafruit's HT1632C library](https://github.com/adafruit/HT1632). 

Feedback welcome :smile: More soon.

### CODE REVIEW

File: lib/led/ledcontrol.js [hunk 1]
```diff
@@ -704,6 +711,128 @@ Controllers = {
...
+      for ( var i = numberOfBits; i > 0; i-- ) {
+        this.io.digitalWrite(this.pins["write"], 0);
+        if ( data & ( 1 << (i-1) ) ) {
+          this.io.digitalWrite(this.pins["data"], 1);
+        } else {
+          this.io.digitalWrite(this.pins["data"], 0);
+        }
+        this.io.digitalWrite(this.pins["write"], 1);
+      }
+      this.io.pinMode(this.pins["data"], this.io.MODES.INPUT);
```
Comment (rwaldron) [inline lib/led/ledcontrol.js]: For all occurrences, just use dot access: `this.pins["write"]` => `this.pins.write`

File: lib/led/ledcontrol.js [hunk 2]
```diff
@@ -704,6 +711,128 @@ Controllers = {
...
+     * @return {[type]} [description]
+     */
+    writeScreen: function() {
+      this.io.digitalWrite(this.pins["cs"], 0);
+      this.writeData(0x5, 3);
+
+      this.writeData(0x0, 7);
+
+      var width = 24, height = 16;
+      for (i=0; i <= (width * height / 8); i+=2) {
```
Comment (rwaldron) [inline lib/led/ledcontrol.js]: `i` is leaking here, needs `var i`

File: lib/led/ledcontrol.js [hunk 3]
```diff
@@ -72,6 +72,13 @@ function LedControl(opts) {
   this.sendDigit = controller.sendDigit;
   this.initialize = controller.initialize;
 
+  // Functions from the HT1632C controller. Perhaps we can find a more declarative
+  // way to copy functions from controller to LEDcontrol (mixin?)
```
Comment (rwaldron) [inline lib/led/ledcontrol.js]: Yes, `LedControl` is a nightmare and I want to toss it for a complete overhaul that follows the Component/Controller model (as seen in just about every other class). Unfortunately, this is a huge undertaking and I don't have the time to do it and no one has stepped forward to tackle it.

File: lib/led/ledcontrol.js [hunk 4]
```diff
@@ -72,6 +72,13 @@ function LedControl(opts) {
...
+  this.sendCommand = controller.sendCommand;
+  this.writeScreen = controller.writeScreen;
+  this.writeData = controller.writeData;
+  this.clearScreen = controller.clearScreen;
```
Comment (rwaldron) [inline lib/led/ledcontrol.js]: This should be defined as `clear`

File: lib/led/ledcontrol.js [hunk 5]
```diff
@@ -704,6 +711,128 @@ Controllers = {
...
+      this.sendCommand(this.OP.LED_ON);
+      this.sendCommand(this.OP.BLINK_OFF);
+      this.sendCommand(this.OP.MASTER_MODE);
+      this.sendCommand(this.OP.INT_RC);
+      // hardcoding this for now. need to figure out a nice way of configuring this.
+      this.sendCommand(this.OP.COMMON_16NMOS);
+      this.sendCommand(this.OP.PWM_CONTROL | 0xF);
+    },
+
+    clearScreen: function() {
```
Comment (rwaldron) [inline lib/led/ledcontrol.js]: `clearScreen` => `clear`

File: lib/led/ledcontrol.js [hunk 6]
```diff
@@ -704,6 +711,128 @@ Controllers = {
...
+      this.pins = {
+        data: opts.pins.data,
+        write: opts.pins.write,
+        cs: opts.pins.cs
+      };
+      ["data", "write", "cs"].forEach(function(pin) {
+        this.io.pinMode(this.pins[pin], this.io.MODES.OUTPUT);
+      }, this);
+
+      this.ledMatrix = [];
```
Comment (rwaldron) [inline lib/led/ledcontrol.js]: This looks like it's serving the same purpose as `this.displaybuffers`?

File: lib/led/ledcontrol.js [hunk 7]
```diff
@@ -704,6 +711,128 @@ Controllers = {
...
+        this.ledMatrix[i] = Math.floor( Math.random() * 256 );
+      }
+      this.writeScreen();
+    },
+
+    /**
+     * Write the data in the buffers to the screen.
+     * @return {[type]} [description]
+     */
+    writeScreen: function() {
```
Comment (rwaldron) [inline lib/led/ledcontrol.js]: This looks like `writeDisplay`. https://github.com/ericandrewlewis/johnny-five/blob/issue-1000/lib/led/ledcontrol.js#L622-L630

File: lib/led/ledcontrol.js [hunk 8]
```diff
@@ -704,6 +711,128 @@ Controllers = {
...
+
+      this.io.digitalWrite(this.pins["cs"], 1);
+    },
+
+    /**
+     * Send a command to the controller.
+     * @param  {Number} commandCode The command code.
+     * @return {[type]}             [description]
+     */
+    sendCommand: function(commandCode) {
```
Comment (rwaldron) [inline lib/led/ledcontrol.js]: `send`

**Comment by ericandrewlewis** [comment]
Thanks for the feedback @rwaldron, I've fixed up accordingly.

Most API functionality has parity with the other matrices.

Here's some example code that would work with an `HT1632C`, hooked up the same way as in [this tutorial](https://learn.adafruit.com/16x24-led-matrix/wiring):

```javascript
var five = require("johnny-five");

var board = new five.Board();

board.on("ready", function() {
  var matrix = new five.Led.Matrix({
    controller: "HT1632C",
    pins: {
      data: 2,
      write: 3,
      cs: 4
    },
    devices: 1,
    dims: {
      rows: 16,
      columns: 24
    }
  });

  matrix.led( 0, 0, true );
  matrix.led( 15, 23, true );
  matrix.writeDisplay();
  matrix.blink();
  matrix.brightness(0);
  // Maximum brightness.
  matrix.brightness(15);
  

  matrix.draw( "hi" );

  matrix.brightness(0);
  matrix.writeDisplay();
});
```

Some open questions:

Writing to the display seems expensive at the moment, so I've disabled `led()` from calling it directly at the moment. Might have to take a look at the C++ library again for optimizations if any are possible. For now, `writeDisplay()` must be called explicitly after `led()`. 

How would one expect `draw()` to work with this display? Passing in an array of row data makes sense. But what about passing in characters, since this display has space for 6 characters?

The HT16K33 supports bicolor displays. Are there bicolor HT1632C matrices?

The HT16K33 supports rotation. Should the HT1632C do this as well?

Any other feedback appreciated.

**Comment by rwaldron** [comment]
> For now, writeDisplay() must be called explicitly after led().

This is a semantic deviation that definitely needs to be resolved, but I have a feeling that won't be possible considering...

> Writing to the display seems expensive at the moment,

Based on the [communication protocol requirements](https://github.com/adafruit/HT1632/blob/master/HT1632.cpp#L402-L441), I assumed it would be. It's not much different from the MAX7219 and MAX722 devices, which are written using `shiftOut` (which is also expensive, though not quite as much)

> How would one expect draw() to work with this display? Passing in an array of row data makes sense. But what about passing in characters, since this display has space for 6 characters?

I honestly have no idea. This display breaks a lot of assumptions made by existing LED matrix component controller implementations—that Johnny-Five just followed because that seemed to be a smart thing to do. Part of me wishes Johnny-Five could offload some of this stuff into plugins like the totally excellent [node-pixel](https://github.com/ajfisher/node-pixel), but that ship has probably sailed :\

Let's gather some other brains that have worked on plugin stuff and/or led component stuff...

@ajfisher @frxnz @boneskull 


The thing I'd like to think about and discuss: building a set of LED component (digit, matrix, bar graph) classes as a plugin similar to node-pixel. I think this would be a way to expand the capabilities in a smart way and most importantly starting _clean_. @dtex and @BrianGenisio are brilliant at identifying capability intersection.

**Comment by boneskull** [comment]
what was the other issue I was talking about this in?  Had something to do with bar graphs...

**Comment by boneskull** [comment]
I think it's worth mentioning that due to the nature of these displays, an inheritance model might be inappropriate. If we define individual capabilities such as "bi-color" or "matrix" and compose resulting objects that way, we won't fall into the trap of trying to classify hardware too hard or violate assumptions. 

I'm not sure a pairing of a public API and a controller object makes much sense though. You can't necessarily use any combination. Maybe the public API should be hardware-specific. I know this goes contrary to much of j5's API but it's worth discussing. The current abstraction for these displays leaks everywhere....

**Comment by boneskull** [comment]
(that ticket was #665)

**Comment by boneskull** [comment]
also, please let me know if the above comment(s) didn't make sense; I wrote them while half-awake

**Comment by ericandrewlewis** [comment]
#665 is a nice read :) 

We could do a refactor there. After that adding support here should be nominal. (Famous last words...) 

I'm going to leave this open for now and see if I can help out on that ticket.

**Comment by dtex** [comment]
Hi @ericandrewlewis ,

Since this has been open and on hold for a while I'm closing. Rather than leave it languishing as an open PR we have created a [Requested Features](../wiki/Requested-Features) page and I have added your request for the HT1632C there with links to this PR and the relevant issue.

**Pull request closed without merging**